### ISSUE TITLE
Support filtering in JavaParserTypeSolver

### ISSUE THREAD

**Issue by jimshowalter** [javaparser/javaparser#2075]
We have some directories with cruft code mixed with actual code. Yes, we could clean that up, but if we could pass a filter to the constructor, then when iterating the directory the solver could skip the files we don't want.

**Comment by ftomassetti** [comment]
I do not think we should do this. It seems very rare that someone would need this

**Comment by jimshowalter** [comment]
The more you don't support stuff like this, the more clients have to roll their own. What is the objection to making the framework tweakable? There's practically no part of, for example, Hibernate that doesn't support delegating in an override.

**Comment by matozoid** [comment]
The main objection is maintenance. Secondary is complexity. The core team has about 1-2 hours a week to spend on JavaParser, and there are 317 bugs to be fixed.

**Comment by jimshowalter** [comment]
That's fair. But in that case, instead of rejecting it, why not put it in a low-priority bucket?

**Comment by matozoid** [comment]
If I look at the type solver, it seems to look only at files ending in ".java" - does that not take care of the problem already?

**Comment by jimshowalter** [comment]
Unfortunately, no. By "cruft", I meant legacy code. Javaparser is great for static analysis of large legacy codebases, but large legacy codebases aren't clean, so acknowledging that and providing a way to filter out unwanted parts of the codebase would be helpful. Otherwise we wind up having to copy the code to a different location and delete unwanted parts of it manually.

You might ask why not just delete the cruft, but even figuring out which files in a large legacy codebase are dead code turns out to be difficult.

**Comment by matozoid** [comment]
We could see if we can replace the custom directory walking logic with standard stuff: https://docs.oracle.com/javase/tutorial/essential/io/walk.html
... and make the FileVisitor class public, and add a parameter to the JavaParserTypeSolver constructor so the visitor can be subclassed and passed in.

If this API is a good fit, that is. But it should give a decent amount of configurability, and improve the code by replacing custom code by library code, so it's a good thing in every way.
